Research/lva-node-app/public/js/mediaGraphBuilder.js

```diff
@@ -0,0 +1,1102 @@
+{
+    let allNodeInfo=[];
+    allNodeInfo.push(getListNodesAndInputs(document.getElementById('dynamic-source')));
+    allNodeInfo.push(getListNodesAndInputs(document.getElementById('dynamic-processor')));
+    allNodeInfo.push(getListNodesAndInputs(document.getElementById('dynamic-sink')));
+    console.log(allNodeInfo);
+    return allNodeInfo;
+}
+
+/**
+ * validates that the graph built by the user doesn't break any rules
+ * @param {any[]} nodesAndInputsList - all nodes and inputs in graph builder
+ * @returns {boolean} if graph is valid or not
+ */
+function validateGraph(nodesAndInputsList)
+{
+    //all processors and sources must have at least one input
+    let referenced=[];
+    referenced=referenced.concat(sourceNames);
+    referenced=referenced.concat(processorNames);
+    if(nodesAndInputsList[2].length==0)
+    {
+        alert("You must have at leastone sink node!");
+        return false;
+    }
+    for(let i=1; i<3; i++) //check processors and sinks for inputs
+    {
+        for(let cur=0; cur<nodesAndInputsList[i].length; cur++)
+        {
+            if(nodesAndInputsList[i][cur].value.length==0)
+            {
+                alert("All processors and sinks must have at least one input. This node has none: " + nodesAndInputsList[i][cur].name);
+                return false;
+            }
+            else
+            {
+                //mark node as having been referenced
+                nodesAndInputsList[i][cur].value.forEach((inputName) => 
+                {
+                    let referencedIndex = referenced.indexOf(inputName);
+                    if(referencedIndex != -1)
+                    {
+                        referenced.splice(referencedIndex, 1);
+                    }
+                })
+            }
+        }
+    }
+    if (referenced.length > 0)
+    {
+        alert("Each node must be referenced by at least one other node. At least this node isn't referenced: "+referenced[0]);
+        return false;
+    }
+    return true;
+}
+
+/** 
+* create a media graph based off of what the user-built 
+* @returns {boolean} - true if successfully created
+*/
+function createMediaGraph() 
+{
+    const graphname = document.getElementById("graphname");
+    const graphDescription = document.getElementById("graph-description");
+    if (graphname.value=="") 
+    {
+        alert("Your graph name cannot be empty!");
+        return;
+    }
+    let sources=[];
+    let processors=[];
+    let sinks=[];
+    let parameters=[]
+    const nodesAndInputs=getGraphNodesAndInputs();
+
+    if(!validateGraph(nodesAndInputs))
+    {
+        return false;
+    }
+    
+    for (let i = 0; i < nodesAndInputs.length; i++) 
+    {
+        for (let node = 0; node < nodesAndInputs[i].length; node++) 
+        {
+            let tempNode = new GraphNode(nodesAndInputs[i][node].name, nodesAndInputs[i][node].value);
+            parameters = parameters.concat(tempNode.nodeParameters);
+            switch (i) 
```

Comment:
assuming sources/processors/sinks be listed in that order and switching on it is super brittle. switch on type or something more explicit.

Reply:
you're right, that was gross. Pushed changes, looks much better now! Using a map instead solves the issue